elpis/engines/hft/objects/model.py

```diff
@@ -0,0 +1,1016 @@
+            cache_dir=self.model_args.cache_dir,
+            activation_dropout=self.model_args.activation_dropout,
+            attention_dropout=self.model_args.attention_dropout,
+            hidden_dropout=self.model_args.hidden_dropout,
+            feat_proj_dropout=self.model_args.feat_proj_dropout,
+            mask_time_prob=self.model_args.mask_time_prob,
+            gradient_checkpointing=self.model_args.gradient_checkpointing,
+            layerdrop=self.model_args.layerdrop,
+            ctc_loss_reduction="mean",
+            pad_token_id=self.processor.tokenizer.pad_token_id,
+            vocab_size=len(self.processor.tokenizer),
+            ctc_zero_infinity=True)
+
+    def preprocess_dataset(self):
+        print("=== Preprocessing Dataset")
+        speech = self.prepare_speech()
+
+        def speech_file_to_array_fn(batch):
+            path = batch['path']
+            start_ms = batch['start_ms']
+            stop_ms = batch['stop_ms']
+            unique_key = f'{path}{start_ms}{stop_ms}'
+            batch["speech"] = speech[unique_key]
+            batch["sampling_rate"] = HFTModel.SAMPLING_RATE
+            batch["target_text"] = batch["text"]
+            batch['duration'] = (batch['stop_ms'] - batch['start_ms'])/1000
+            batch['duration'] = len(batch['speech'])/batch['sampling_rate']
+            return batch
+
+        self.hft_dataset = self.hft_dataset.map(
+            speech_file_to_array_fn,
+            remove_columns=self.hft_dataset['train'].column_names,
+            num_proc=self.data_args.preprocessing_num_workers,
+        )
+        print("=== hft_dataset")
+        print(self.hft_dataset)
+
+
+    def prepare_dataset(self):
+        print("=== Preparing Dataset")
+        def prepare_dataset(batch):
+            # Check that all files have the correct sampling rate
+            assert (
+                len(set(batch["sampling_rate"])) == 1
+            ), f"Make sure all inputs have the same sampling rate of {self.processor.feature_extractor.sampling_rate}."
+            batch["input_values"] = self.processor(batch["speech"], sampling_rate=batch["sampling_rate"][0]).input_values
+            # Setup the processor for targets
+            with self.processor.as_target_processor():
+                batch["labels"] = self.processor(batch["target_text"]).input_ids
+            return batch
+
+        self.hft_dataset = self.hft_dataset.map(
+            prepare_dataset,
+            remove_columns=self.hft_dataset['train'].column_names,
+            batch_size=self.training_args.per_device_train_batch_size,
+            batched=True,
+            num_proc=self.data_args.preprocessing_num_workers,
+        )
+
+    def prepare_speech(self):
+        print("=== Preparing Speech")
+        speech = {}
+        audio_paths = set()
+        rejected_count = 0
+        for utt in self.hft_dataset['train']:
+            audio_paths.add((utt['path'], utt['text'], utt['start_ms'], utt['stop_ms']))
```

Comment:
Function?

Reply:
@mattchrlw your call..

Reply:
Sorry, could you clarify @nicklambourne?